src/renderer/shared/api/governance/on-chain/service/claimScheduleService.ts

```diff
@@ -0,0 +1,462 @@
+import { BN, BN_ZERO } from '@polkadot/util';
+import uniqWith from 'lodash/uniqWith';
+import isEqual from 'lodash/isEqual';
+import orderBy from 'lodash/orderBy';
+import sortBy from 'lodash/sortBy';
+import clone from 'lodash/clone';
+
+import { onChainUtils } from '../lib/on-chain-utils';
+import {
+  type ClaimableLock,
+  AffectVote,
+  AffectTrack,
+  ClaimTime,
+  ClaimTimeAt,
+  GroupedClaimAffects,
+  ClaimAffect,
+  ClaimAction,
+  Unlock,
+  RemoveVote,
+  UnlockChunk,
+  PendingChunk,
+  ClaimableChunk,
+  ClaimTimeUntil,
+} from '../lib/types';
+import type {
+  BlockHeight,
+  ReferendumInfo,
+  TrackInfo,
+  Voting,
+  ReferendumId,
+  TrackId,
+  CastingVoting,
+  AccountVote,
+  TimedOutReferendum,
+  OngoingReferendum,
+  StandardVote,
+  DelegatingVoting,
+} from '@shared/core';
+
+export const claimScheduleService = {
+  estimateClaimSchedule,
+};
+
+type ClaimParams = {
+  currentBlockNumber: BlockHeight;
+  referendums: Record<ReferendumId, ReferendumInfo>;
+  tracks: Record<TrackId, TrackInfo>;
+  trackLocks: Record<TrackId, BN>;
+  votingByTrack: Record<TrackId, Voting>;
+  undecidingTimeout: BlockHeight;
+  voteLockingPeriod: BlockHeight;
+};
+function estimateClaimSchedule({
+  currentBlockNumber,
+  referendums,
+  tracks,
+  trackLocks,
+  votingByTrack,
+  undecidingTimeout,
+  voteLockingPeriod,
+}: ClaimParams): UnlockChunk[] {
+  // step 1 - determine/estimate individual unlocks for all priors and votes
+  // result example: [(1500, 1 KSM), (1200, 2 KSM), (1000, 1 KSM)]
+  const claimableLocks = individualClaimableLocks(
+    currentBlockNumber,
+    referendums,
+    votingByTrack,
+    tracks,
+    trackLocks,
+    voteLockingPeriod,
+    undecidingTimeout,
+  );
+
+  // step 2 - fold all locks with same lockAt
+  // { 1500: 1 KSM, 1200: 2 KSM, 1000: 1 KSM }
+  const maxUnlockedByTime = combineSameUnlockAt(claimableLocks);
+
+  // step 3 - convert individual schedule to global
+  // [(1500, 1 KSM), (1200, 1 KSM)]
+  const unlockSchedule = constructUnlockSchedule(maxUnlockedByTime);
+
+  // step 4 - convert locks affects to claim actions
+  return toUnlockChunks(unlockSchedule, currentBlockNumber);
+}
+
+// Step 1
+function individualClaimableLocks(
```

Comment:
Some functions in this file called as variables, are you sure about such naming?

Reply:
Named it as close as possible to Kotlin's code to easier compare with original code if bugs happen